### PR TITLE
Added propcore.

### PR DESCRIPTION
**Pull request by thegrb93**
Let me know if anything is wrong. I've tested it and it works.

### CODE REVIEW

**Comment by awilliamson** [comment]
Codestyle. First thing you should check when submitting a PR, it's all on the readme.

**Comment by awilliamson** [comment]
Also @thegrb93, The reason issue #19 was Entity/Propcore was because Propcore is a special case of an entity spawning system. This should really be done by users specifying prop_physics as the entity.

But as previously mentioned, it's a permissions nightmare and open to significant abuse. If something can be written by the users, then it should be. We're not about adding convenience everywhere when they have all the tools to implement it. ( In this case that would be through the yet-to-be-completed entity core ).

Can @Xandaros chime in on this, just to confirm or reject any statements. Want to make sure we're on the same page.

**Comment by thegrb93** [comment]
Mind putting comments where my mistakes are? I don't see any.

**Comment by thegrb93** [comment]
Why over-complicate when all you need is prop_physics? I guess you could do a white list if you wanted to allow people to be able to spawn other entities, but that doesn't make much sense imo.

**Comment by thegrb93** [comment]
Removed the duplicator functions. I was just working off of Garry's MakeProp code.

**Comment by awilliamson** [comment]
```
Mind putting comments where my mistakes are? I don't see any.
```

Codestyle everywhere, almost all function declarations, table indexing etc. It might just be easy for me to spot because we've been doing it for so long. Make sure you carefully look at the codestyle guidelines and the additions in your commit.

```
Why over-complicate when all you need is prop_physics?
```

The Entity core is still needed, and props are a sub-set of entities; Specifically, prop_physics. All props are entities, but not all entities are props. The 'whitelist' you mention is the entire permissions system, and it very much does make sense. Want admins to spawn some entities, but not regular players: permissions. Want to give general players the ability to spawn custom entities you've added with starfall: permissions.

What doesn't make sense is implementing a prop core system which is pretty much copy/paste of a single-use case of the larger entity system. This is one of the reasons that this system has been on hold, as it requires careful design decisions. We really wish to avoid this 'idea' that convenience is 'key' to the users, and we've opted for the more 'proper' implementation of a system. The tools are provided for the users, therefore they can make these systems if they don't require specific interfacing with GLua. There are some exceptions to this rule, but generally it holds.

Please wait for @Xandaros to comment on this, he might be able to explain our stance a bit more clearly than I can.

**Comment by thegrb93** [comment]
The coding style is the same as the hologram library so you might want to fix the hologram library's style.

**Comment by awilliamson** [comment]
See #7. Changing an existing codebase over to new codestyle guidelines is 'difficult' work. You were told multiple times what guidelines to follow and where to find them. Those guidelines hold for any PR, including those by @Xandaros and myself. As we change sections of the codebase we oftentimes fix codestyle of surrounding sections as we go, thus eventually ensuring all code conforms to the guidelines.

Now please, this is the last time I will tell you, look at the codestyle guidelines in the readme and follow them. Your PR will **_not_** be accepted if it does not conform to the stated guidelines.

**Comment by thegrb93** [comment]
I don't have time or patience to mess with that. I'm offering some help with speeding up the development of needed libraries, but the code stylizing is totally unnecessary imo. You can do the stylizing yourself if you want it, but I'll be hosting my improvements on my repository so it doesn't matter to me whether or not you pull it.

**Comment by wraith321** [comment]
omg ^^

File: lua/starfall/libs_sv/prop.lua [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	if not can_spawn( instance ) then return SF.throw( "Can't spawn props that often", 2 )
+	elseif personal_max_reached( instance ) then return SF.throw( "Can't spawn props, maximum personal limit of " .. SF.Props.personalquota:GetInt() .. " has been reached", 2 )
+	elseif max_reached() then return SF.throw( "Can't spawn props, maximum limit of " .. SF.Props.defaultquota:GetInt() .. " has been reached", 2 ) end
+	if not IsValid( instance.player ) then return SF.Entities.Wrap( NULL ) end
+	if not gamemode.Call( "PlayerSpawnProp", instance.player, model ) then return SF.Entities.Wrap( NULL ) end
+
+	local propdata = instance.data.props
+	local propent = ents.Create( "prop_physics" )
+	
+	propent:CallOnRemove( "starfall_propgram_delete", propOnDestroy, propdata )
```
Comment (awilliamson) [inline lua/starfall/libs_sv/prop.lua]: Nice 'propgram' you have here. Shame if this entire PR was a copy/paste of my hologram library, just reworded slightly.

**Comment by awilliamson** [comment]
``` lua
propent:CallOnRemove( "starfall_propgram_delete", propOnDestroy, propdata, instance.player )
```

Clear evidence of find/replace mistake, therefore evidence of copy/paste. Closing this PR as lack of brain cells.

@thegrb93 It's very ignorant and brazen to brandish work, claiming it to be your own, to an upstream repository without checking it, showing resistance to PR guidelines when stated, and providing code which is a blatant copy-paste of our existing work. Yes the systems are similar, but copy-pasting without understanding why breeds bad coders. I will not help reinforce a model that promotes plagarism and lack of originality and application of knowledge.

**Comment by thegrb93** [comment]
It's funny you think reusing open code (in the same project no less) is considered plagiarism. I copy pasted it to save time and got the task done. Something you still haven't even started after a year. Thanks for spotting that name I missed I guess, even though it doesn't affect the program at all. I don't have any obligation to follow your stupid guidelines. You can accept my help and edit my code to whatever standard you want, but it looks like you would rather not even have a prop core.

**Comment by Xandaros** [comment]
We'd rather have missing features than unmaintainable code.
Granted, our standards are quite high and the original codebase still needs to be put up to standard, but new features should follow the guidelines to make it easier to maintain the code and have a common style across the whole project.
Fixing the code style shouldn't be too much of an issue, as most of the copied code is already in the correct style, but there are still a few places that don't follow it.

You might also want to either squash the commits or change the commit messages to more meaningful names. "Added permissions"; for example, makes sense in the context of the PR, but once it is merged in, this context is lost and from the message alone, you might think that it adds the whole permission system.

As for copying from the holograms library, that's fine, but you might want to mention it in the commit message. Not doing so is indeed plagiarism. (Yes, it is open source, but if you read the license you'd see that you are required to credit the respective author. Since it is in the same project, just saying where you got it from is sufficient, since authorship is implicit in the git history)

**Pull request closed without merging**